frontend/src/metabase/lib/expressions/typeinferencer.js

```diff
@@ -0,0 +1,54 @@
+import { MBQL_CLAUSES } from "./config";
+import { OPERATOR as OP } from "./tokenizer";
+
+export const MONOTYPE = {
+  Undefined: "undefined",
+  Number: "number",
+  String: "string",
+  Boolean: "boolean",
+};
+
+export function infer(mbql) {
+  if (!Array.isArray(mbql)) {
+    return typeof mbql;
+  }
+  const op = mbql[0];
+  switch (op) {
+    case OP.Plus:
+    case OP.Minus:
+    case OP.Star:
+    case OP.Slash:
+      return MONOTYPE.Number;
+
+    case OP.Not:
+    case OP.And:
+    case OP.Or:
+    case OP.Equal:
+    case OP.NotEqual:
+    case OP.GreaterThan:
+    case OP.GreaterThanEqual:
+    case OP.LessThan:
+    case OP.LessThanEqual:
+      return MONOTYPE.Boolean;
+  }
+
+  if (op === "case" || op === "coalesce") {
```

Comment:
Would `concat` also be a possible mixed type? Issue 12938

Reply:
It shouldn't be, as `concat` should always return a string. That issue is about lack of implicit casting (which needs to be solved, but tangent to this).